src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -687,3 +688,145 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
+      if (parentPartitionBoundaries(p) > firstRowInNextBlock)
+        p -= 1
+      
+      a(blockRowIndex) = (start, skip, end)
+      
+      firstRowInBlock = firstRowInNextBlock
+      blockRowIndex += 1
+    }
+    
+    a
+  }
+}
+
+// IRM must be complete (IRM numRows == RDD count) and ordered (IndexedRow index == RDD index); checked by assertions
+class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, blockSize: Int) extends RDD[Int](irm.rows.sparkContext, Nil) {
+  private val rows = irm.numRows()
+  private val cols = irm.numCols()
+  private val parentPartitions = irm.rows.partitions
+  private val parentPartitionBoundaries: Array[Long] = irm.rows.countPerPartition().scanLeft(0L)(_ + _)
+  
+  assert(rows == parentPartitionBoundaries.last,
+    s"IndexedRowMatrix has $rows rows but RDD only has ${parentPartitionBoundaries.last} IndexedRows.")
+  
+  private val gp = GridPartitioner(blockSize, rows, cols)
+  private val truncatedBlockRow = rows / blockSize
+  private val truncatedBlockCol = cols / blockSize
+  private val excessRows = (rows % blockSize).toInt
+  private val excessCols = (cols % blockSize).toInt
+  
+  private val d = digitsNeeded(gp.numPartitions)
+  private val sHadoopBc = irm.rows.sparkContext.broadcast(
+    new SerializableHadoopConfiguration(irm.rows.sparkContext.hadoopConfiguration))
+
+  override def getDependencies: Seq[Dependency[_]] = {  
+    Array[Dependency[_]](
+      new NarrowDependency(irm.rows) {
+        def getParents(partitionId: Int): Seq[Int] =
+          partitions(partitionId).asInstanceOf[WriteBlocksRDDPartition].range
+      }
+    )
+  }
+  
+  protected def getPartitions: Array[Partition] = {
+    WriteBlocksRDD.computeBlockRowDependencies(parentPartitionBoundaries, gp)
+      .zipWithIndex
+      .map { case ((start, skip, end), blockRowIndex) =>
+          WriteBlocksRDDPartition(blockRowIndex, start, skip, end)
+      }
+  }
+  
+  def compute(split: Partition, context: TaskContext): Iterator[Int] = {
+    val blockRowIndex = split.index
+    val firstRowInBlock = blockRowIndex.toLong * blockSize
+    val nRowsInBlock = if (blockRowIndex != truncatedBlockRow) blockSize else excessRows
+
+    val dosArray = Array.tabulate(gp.colPartitions) { blockColIndex =>
+      val nColsInBlock = if (blockColIndex == truncatedBlockCol) excessCols else blockSize
+
+      val is = gp.partitionIdFromBlockIndices(blockRowIndex, blockColIndex).toString
+      assert(is.length <= d)
+      val pis = StringUtils.leftPad(is, d, "0")
+      val filename = path + "/parts/part-" + pis
+
+      val dos = new DataOutputStream(sHadoopBc.value.value.unsafeWriter(filename))      
+      dos.writeInt(nRowsInBlock)
+      dos.writeInt(nColsInBlock)
+      dos.writeBoolean(true) // transposed, stored row major
+      
+      dos
+    }
+    
+    val split0 = split.asInstanceOf[WriteBlocksRDDPartition]
+    val start = split0.start
+    var i = -split0.skip
```

Comment:
Too clever by half.  Just call `next()` `skip` times in the `p == start` conditional.